Compile item group not populated without warning message

### Issue Description

I was defining multiple custom properties and items containing files above the project, so outside its scope, in order to include them in the NuGet package (ReadMe, Citation, documentation files).
I was doing that early in a props file.
The evaluation phase of MSBuild then decides that the project may not use `EnableDefaultCompileItems` and forces it to `false`.

### Steps to Reproduce

Clone [this repo](https://github.com/Scal-Human/DotNet.MSBuild.Issue.13566).
```
git clone https://github.com/Scal-Human/DotNet.MSBuild.Issue.13566
cd DotNet.MSBuild.Issue.13566\Source\SomeProject
dotnet build
```
Projects do not compile as the `Compile` item is empty because the `EnableDefaultCompileItems` is `false` while there is no apparent relationship between this property and the default compilation items.
Output:
```
  Determining projects to restore...
  ProjectBeforeRestore: EnableDefaultCompileItems false
  ProjectBeforeRestore: EnableDefaultItems true
...\Source\SomeProject\SomeProject.csproj(15,9): warning : ProjectBeforeRestore: Compile ItemGroup is empty
  All projects are up-to-date for restore.
CSC : error CS5001: Program does not contain a static 'Main' method suitable for an entry point [C:\Scal26\DotNet.MSBuild.Issue.13566\Source\SomeProject\SomeProject.csproj]

Build FAILED.
```

### Expected Behavior

MSBuild or the Sdk should emit at least a warning instead of losing the time of the user searching for apparently unrelated side-effect.


### Actual Behavior

The bug is to do that **without informing the user** of such a radical decision.


### Analysis

Projects do not compile anymore as the `Compile` item is empty because the `EnableDefaultCompileItems` is `false` while there is no apparent relationship between this property and the default compilation items.
> Note that forcing  `EnableDefaultCompileItems` to `true`is simply ignored.

### Workaround
Such files outside the scope of the project must be handled in a target; i.e. after the evaluation phase.
In the case above a target defining those properties and items executed before the `Pack` does the job.

### Versions & Configurations

MSBuild 18.0.11.7603
Sdk 10.0.103

@Scal-Human
>making it simpler is removing the problem.

That is the point: it does not. It fails with the same error as your project.
>This error occurs because something happens before the import of Microsoft.NET.Sdk.

Do you have any evidence to support this conclusion? Does it build without error when you move
```xml
<Import Project="Sdk.props"     Sdk="Microsoft.NET.Sdk" />
```
on top of `Sdk.props`? For me it fails with the same error.

>I repeat: the actual bug is that we are not informed (info, warning or whatever) of a decision.

I can repeat too: do you have any evidence to support this conclusion? If you are so unwilling to inspect value of `EnableDefaultCompileItems` on build run yourself, then let me do it for you:
```xml
<Project Sdk="$(MSBuildThisFileDirectory)..\CustomSdk">

    <PropertyGroup>
        <OutputType>Exe</OutputType>
        <TargetFramework>net10.0</TargetFramework>
        <ImplicitUsings>enable</ImplicitUsings>
        <Nullable>enable</Nullable>
    </PropertyGroup>

    <Target Name="ProjectBeforeRestore" BeforeTargets="Restore">
        <Message Importance="High"  Text="ProjectBeforeRestore: EnableDefaultCompileItems $(EnableDefaultCompileItems)" />
        <Message Importance="High"  Text="ProjectBeforeRestore: EnableDefaultItems $(EnableDefaultItems)" />
        <Warning                    Text="ProjectBeforeRestore: Compile ItemGroup is empty"   Condition="'@(Compile)' == ''" />
        <Message Importance="High"  Text="ProjectBeforeRestore: Compile %(Compile.Identity)"  Condition="'@(Compile)' != ''" />
    </Target>

    <Target Name="InspectEnableDefaultCompileItems" BeforeTargets="Restore; BeforeBuild">
        <Warning Text="InspectEnableDefaultCompileItems: $(EnableDefaultCompileItems)" />
    </Target>

</Project>
```
Results:
```
Restore succeeded with 2 warning(s) in 0,6s
    M:\Test\Source\SomeProject\SomeProject.csproj(13,9): warning ProjectBeforeRestore: Compile ItemGroup is empty
    M:\Test\Source\SomeProject\SomeProject.csproj(18,9): warning InspectEnableDefaultCompileItems: false
  SomeProject net10.0 failed with 1 error(s) and 1 warning(s) (0,3s)
    M:\Test\Source\SomeProject\SomeProject.csproj(18,9): warning InspectEnableDefaultCompileItems: true
    CSC : error CS5001: Program does not contain a static 'Main' method suitable for an entry point

Build failed with 1 error(s) and 3 warning(s) in 1,6s
```
Wow. `EnableDefaultCompileItems` is actually true on build run.
>`Build` will not run

Yes, it would not (that is why you need to use `BeforeTargets="BeforeBuild"` here), but for very different reasons, than you describe.

--- a/src/Build.UnitTests/Definition/ProjectEvaluationContext_Tests.cs
+++ b/src/Build.UnitTests/Definition/ProjectEvaluationContext_Tests.cs
@@ -13,6 +13,7 @@ using Microsoft.Build.Evaluation;
 using Microsoft.Build.Evaluation.Context;
 using Microsoft.Build.Execution;
 using Microsoft.Build.Framework;
+using Microsoft.Build.Shared.FileSystem;
 using Microsoft.Build.Unittest;
 using Shouldly;
 using Xunit;
@@ -894,6 +895,53 @@ namespace Microsoft.Build.UnitTests.Definition
                 });
 
             evaluationCount.ShouldBe(2);
+        }
+
+        [Fact]
+        public void CachingFileSystemWrapperDistinguishesFileAndDirectoryExistence()
+        {
+            var directory = _env.DefaultTestDirectory.CreateDirectory("subDirectory");
+            var file = _env.DefaultTestDirectory.CreateFile("file");
+            var fileSystem = new CachingFileSystemWrapper(FileSystems.Default);
+
+            fileSystem.FileExists(directory.Path).ShouldBeFalse();
+            fileSystem.DirectoryExists(directory.Path).ShouldBeTrue();
+            fileSystem.FileOrDirectoryExists(directory.Path).ShouldBeTrue();
+
+            fileSystem.DirectoryExists(file.Path).ShouldBeFalse();
+            fileSystem.FileExists(file.Path).ShouldBeTrue();
+            fileSystem.FileOrDirectoryExists(file.Path).ShouldBeTrue();
+        }
+
+        [Theory]
+        [InlineData(EvaluationContext.SharingPolicy.Isolated)]
+        [InlineData(EvaluationContext.SharingPolicy.SharedSDKCache)]
+        [InlineData(EvaluationContext.SharingPolicy.Shared)]
+        public void GetDirectoryNameOfFileAboveWithEmptyFileNameDoesNotPoisonProjectLevelWildcards(EvaluationContext.SharingPolicy policy)
+        {
+            var context = EvaluationContext.Create(policy);
+            _env.DefaultTestDirectory.CreateFile("Source.cs");
+
+            EvaluateProjects(
+                new[]
+                {
+                    """
+                    <Project>
+                      <PropertyGroup>
+                        <SearchedPath>$([MSBuild]::GetDirectoryNameOfFileAbove('$(MSBuildProjectDirectory)', ''))</SearchedPath>
+                      </PropertyGroup>
+                      <ItemGroup>
+                        <i Include="*.cs" />
+                      </ItemGroup>
+                    </Project>
+                    """
+                },
+                context,
+                project =>
+                {
+                    project.GetPropertyValue("SearchedPath").ShouldBeEmpty();
+                    ObjectModelHelpers.AssertItems(["Source.cs"], project.GetItems("i"));
+                });
         }
 
         [Theory]

--- a/src/Framework/FileSystem/CachingFileSystemWrapper.cs
+++ b/src/Framework/FileSystem/CachingFileSystemWrapper.cs
@@ -12,10 +12,10 @@ namespace Microsoft.Build.Shared.FileSystem
     internal sealed class CachingFileSystemWrapper : IFileSystem
     {
         private readonly IFileSystem _fileSystem;
-        private readonly ConcurrentDictionary<string, bool> _directoryExistenceCache = new ConcurrentDictionary<string, bool>();
-        private readonly ConcurrentDictionary<string, bool> _fileExistenceCache = new ConcurrentDictionary<string, bool>();
-        private readonly ConcurrentDictionary<string, bool> _fileOrDirectoryExistenceCache = new ConcurrentDictionary<string, bool>();
-        private readonly ConcurrentDictionary<string, DateTime> _lastWriteTimeCache = new ConcurrentDictionary<string, DateTime>();
+        private readonly ConcurrentDictionary<string, bool> _directoryExistenceCache = new();
+        private readonly ConcurrentDictionary<string, bool> _fileExistenceCache = new();
+        private readonly ConcurrentDictionary<string, bool> _fileOrDirectoryExistenceCache = new();
+        private readonly ConcurrentDictionary<string, DateTime> _lastWriteTimeCache = new();
 
         public CachingFileSystemWrapper(IFileSystem fileSystem)
         {
@@ -24,6 +24,13 @@ namespace Microsoft.Build.Shared.FileSystem
 
         public bool FileOrDirectoryExists(string path)
         {
+            // A positive result from either specific cache implies existence, so we can avoid a redundant filesystem stat.
+            if ((_fileExistenceCache.TryGetValue(path, out bool fileExists) && fileExists) ||
+                (_directoryExistenceCache.TryGetValue(path, out bool directoryExists) && directoryExists))
+            {
+                return true;
+            }
+
             return _fileOrDirectoryExistenceCache.GetOrAdd(path, p => _fileSystem.FileOrDirectoryExists(p));
         }
 

--- a/src/Framework/FileSystem/CachingFileSystemWrapper.cs
+++ b/src/Framework/FileSystem/CachingFileSystemWrapper.cs
@@ -12,7 +12,9 @@ namespace Microsoft.Build.Shared.FileSystem
     internal sealed class CachingFileSystemWrapper : IFileSystem
     {
         private readonly IFileSystem _fileSystem;
-        private readonly ConcurrentDictionary<string, bool> _existenceCache = new ConcurrentDictionary<string, bool>();
+        private readonly ConcurrentDictionary<string, bool> _directoryExistenceCache = new ConcurrentDictionary<string, bool>();
+        private readonly ConcurrentDictionary<string, bool> _fileExistenceCache = new ConcurrentDictionary<string, bool>();
+        private readonly ConcurrentDictionary<string, bool> _fileOrDirectoryExistenceCache = new ConcurrentDictionary<string, bool>();
         private readonly ConcurrentDictionary<string, DateTime> _lastWriteTimeCache = new ConcurrentDictionary<string, DateTime>();
 
         public CachingFileSystemWrapper(IFileSystem fileSystem)
@@ -22,7 +24,7 @@ namespace Microsoft.Build.Shared.FileSystem
 
         public bool FileOrDirectoryExists(string path)
         {
-            return CachedExistenceCheck(path, p => _fileSystem.FileOrDirectoryExists(p));
+            return _fileOrDirectoryExistenceCache.GetOrAdd(path, p => _fileSystem.FileOrDirectoryExists(p));
         }
 
         public FileAttributes GetAttributes(string path)
@@ -37,12 +39,12 @@ namespace Microsoft.Build.Shared.FileSystem
 
         public bool DirectoryExists(string path)
         {
-            return CachedExistenceCheck(path, p => _fileSystem.DirectoryExists(p));
+            return _directoryExistenceCache.GetOrAdd(path, p => _fileSystem.DirectoryExists(p));
         }
 
         public bool FileExists(string path)
         {
-            return CachedExistenceCheck(path, p => _fileSystem.FileExists(p));
+            return _fileExistenceCache.GetOrAdd(path, p => _fileSystem.FileExists(p));
         }
 
         public IEnumerable<string> EnumerateDirectories(string path, string searchPattern = "*", SearchOption searchOption = SearchOption.TopDirectoryOnly)
@@ -78,11 +80,6 @@ namespace Microsoft.Build.Shared.FileSystem
         public IEnumerable<string> EnumerateFileSystemEntries(string path, string searchPattern = "*", SearchOption searchOption = SearchOption.TopDirectoryOnly)
         {
             return _fileSystem.EnumerateFileSystemEntries(path, searchPattern, searchOption);
-        }
-
-        private bool CachedExistenceCheck(string path, Func<string, bool> existenceCheck)
-        {
-            return _existenceCache.GetOrAdd(path, existenceCheck);
         }
     }
 }

